Helper/Order.php

```diff
@@ -1336,8 +1336,19 @@ public function setOrderState($order, $state)
                 $order->setStatus($order->getConfig()->getStateDefaultStatus(OrderModel::STATE_HOLDED));
             }
         } elseif ($state == OrderModel::STATE_CANCELED) {
-            $this->cancelOrder($order);
-            return;
+            if ($order->canCancel()){
+                $this->cancelOrder($order);
+                return;
```

Comment:
Just for my understanding:

What is the difference between cancelling order here and the registerCancel below. Why can't we use the below one for all cases

Reply:
The difference here is that we’re dealing with 2 separate cases for canceling the order:

1. Normal order cancelation (e.g. void hook request)

2. Canceling the order when the payment is rejected irreversibly (payment is denied)

For #1, Magento cancels the payment and then registers cancelation. See here https://screencast.com/t/c5OHnLfv5lC

For #2, Magento just registers cancelation

This is how default Magento works, we can’t use #1 for #2 as the order state is payment review and we can’t cancel the order in that case https://screencast.com/t/KwjNEKdPQ3Y

We can use #2 for #1, but the logic to cancel the payment won’t be run and I’m not sure if it will cause any issues. I think it’s better to keep the logic like it is here. Thoughts?